### PR TITLE
expat: 2.4.4 -> 2.4.5

### PR DESCRIPTION
**Pull request by hartwork**
###### Motivation for this change
[libexpat 2.4.5](https://github.com/libexpat/libexpat/releases/tag/R_2_4_5) with **security fixes** has been released, the upstream [change log](https://github.com/libexpat/libexpat/blob/R_2_4_5/expat/Changes) has more details.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bobby285271** [comment]
@ofborg eval

**Review by vcunat — APPROVED** [review]
The changelog seems quite focused just on security fixes, so I assume that risk of some breakage is rather low. I tested building some packages. I see no use in waiting more.

**Comment by risicle** [comment]
Perhaps we'd be better off spending our testing effort on 21.11...

**Comment by hartwork** [comment]
> Perhaps we'd be better off spending our testing effort on 21.11...

Could you elaborate?

**Comment by risicle** [comment]
We're doing the bump in `staging-21.11` too, and it's possible that it'll hit real users there before it makes it in to `master`. So I'll be doing some testing on #160897. That's all.

**Comment by trofi** [comment]
Bisect says `perlPackages.XMLParser` is broken by 62b1a57752816207137e3f7f8c1167c80eb08965 `expat: 2.4.4 -> 2.4.5 (security)`:

```
$ nix build -f. perlPackages.XMLParser -L
...
perl5.34.0-XML-Parser> #   Failed test at t/decl.t line 86.
perl5.34.0-XML-Parser> #          got: '((bar|foo|xyz+),bar)'
perl5.34.0-XML-Parser> #     expected: '((bar|foo|xyz+),zebra*)'
perl5.34.0-XML-Parser> #   Failed test at t/decl.t line 95.
perl5.34.0-XML-Parser> #          got: 'bar'
perl5.34.0-XML-Parser> #     expected: 'zebra'
perl5.34.0-XML-Parser> #   Failed test at t/decl.t line 96.
perl5.34.0-XML-Parser> #          got: undef
perl5.34.0-XML-Parser> #     expected: '*'
perl5.34.0-XML-Parser> # Looks like you failed 3 tests of 40.
```

**Comment by hartwork** [comment]
@trofi can you help me find that test case so I can debug this in Expat?

**Comment by risicle** [comment]
https://metacpan.org/release/TODDR/XML-Parser-2.46/source/t/decl.t#L86

**Comment by hartwork** [comment]
Thanks!

**Comment by SuperSandro2000** [comment]
Next time it would be better to debug this first instead of leaving staging in a totally broken state.

**Comment by SuperSandro2000** [comment]
also @hartwork you might want to add yourself to meta.maintainers.

PS: Why do we not have an extra tag for upstream developers which also maintain the package?

**Comment by hartwork** [comment]
> also @hartwork you might want to add yourself to meta.maintainers.

@SuperSandro2000 I'm not really maintaining the package and I have enough to do upstream already. I originally jumped in with Expat package updates at #124212 because Expat was badly out of date in nixOS at the time. Maybe it's best if I take my hands back off this package and someone else takes over.

**Comment by vcunat** [comment]
I don't think staging is totally broken by this.  And I did test `libxml2` (and the stdenv rebuild) before merging.

**Comment by risicle** [comment]
Seconded - I've been building many, many packages without issue.

**Comment by SuperSandro2000** [comment]
`perlPackages.XMLParser` is a dependency of nix and the ofborg-eval-lib-tests currently fail because of that which means every PR against staging is red.

**Comment by SuperSandro2000** [comment]
> Maybe it's best if I take my hands back off this package and someone else takes over.

Not your fault that `perlPackages.XMLParser` is broken.

---

Anyone has any idea how to fix it? Is it just the test or code used by other things? Otherwise we need to revert this update until someone can fix it because right now every PR is red and normally can't be merged and staging would never advance.

**Comment by hartwork** [comment]
> Anyone has any idea how to fix it? Is it just the test or code used by other things?

I'm at it. I'll need more time.

**Comment by risicle** [comment]
Mental note: this is the sort of case where adding `perlPackages.XMLParser` to `expat`'s `passthru.tests` would help us make sure it gets tested next time.

**Comment by SuperSandro2000** [comment]
> Mental note: this is the sort of case where adding `perlPackages.XMLParser` to `expat`'s `passthru.tests` would help us make sure it gets tested next time.

image

**Comment by hartwork** [comment]
> I'm at it. I'll need more time.

The regression turned out real, there is a fix, a new regression test, a new release 2.4.6 upstream, and a NixOS pull request #161090 to bump to 2.4.6.

**Comment by trofi** [comment]
Thank you, Sebastian!

**Comment by hartwork** [comment]
@trofi thanks for the report!

**Comment by risicle** [comment]
Would you like us to set up a full nixpkgs rebuild as your expat project CI? It would only take a few weeks per commit... :trollface:

**Comment by hartwork** [comment]
@risicle I think I need to respectfully decline :smiley: (If there is a serious bit in there, let's take it to e-mail please.)

**Pull request merged**
